railgun/hack/generators/manifests/main.go

```diff
@@ -0,0 +1,124 @@
+// but not a valid part of the API. In v1 this field is effectively deprecated
+// and MUST be set to false:
+//
+//  https://github.com/kubernetes/kubernetes/pull/93078
+//
+// Problematically: even though validation enforces the "false" setting of this
+// deprecated attribute for v1 CRDs, the previous default was "true" in v1beta1
+// and if you try to apply a v1 CRD that does NOT have this field set where a
+// v1beta1 version of the CRD already exists on the cluster this can result in
+// and error and fail to upgrade the CRD.
+//
+// This program accounts for this by applying "preserveUnknownFields: false" to
+// any v1 CustomResourceDefinition to work around the upstream backwards
+// incompatibility issue automatically.
+//
+// NOTE: When this was first written using controller-gen configuration options
+//       and kubebuilder flags on the API types was attempted but none of these
+//       things could produce the right result.
+//
+//       The controller-gen CLI can't really do it (easily) because the default
+//       behavior of the marshaller in sigs.k8s.io/yaml is to not emit optional
+//       boolean fields to YAML if they're set to false. 😑
+//
+//       It's quite possible there's a better way to do this and if you find one
+//       we encourage you to put in a PR and replace this.
+//
+// -----------------------------------------------------------------------------
+
+func main() {
+	// handle arguments
+	flag.StringVar(&directory, "directory", ".", "which directory to find CRDs in")
+	flag.Parse()
+
+	// find all the YAML CRDs in the provided directory and patch them
+	if err := filepath.Walk(directory, processFile); err != nil {
+		fmt.Fprintf(os.Stderr, "Error processing files in %s: %s\n", directory, err)
+		os.Exit(1)
+	}
+}
+
+func processFile(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if info.IsDir() {
+		return nil
+	}
+
+	// filter out any files that don't identify themselves as YAML files via file extensions
+	if !strings.HasSuffix(info.Name(), ".yaml") && !strings.HasSuffix(info.Name(), ".yml") {
+		return nil
+	}
+
+	// read in the file contents
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	// filter out any YAML files that don't contain a v1.CustomResourceDefinition
+	if !bytes.Contains(b, []byte(`kind: CustomResourceDefinition`)) || !bytes.Contains(b, []byte(`apiVersion: apiextensions.k8s.io/v1`)) {
+		return nil
+	}
+
+	// don't allow files that contain multiple objects (controller-gen wont be configured to emit these anyway)
+	yamlCRDs := bytes.Split(b, []byte(`\n---\n`))
+	if len(yamlCRDs) > 1 {
+		return fmt.Errorf("file %s contained multiple objects (%d) which this script doesn't yet support", path, len(yamlCRDs))
+	}
+
+	// generate a patch command to ensure "preserveUnknownFields" is set to false
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
```

Comment:
dont check stderr ?

Reply:
`stderr` and `err` are basically mutually exclusive. If an error occurs we catch it below in `err` and then we print out any captured stderr output in the error.

Reply:
Line 107 presents stderr, it'd better check error in this case, and this error is thrown by new function.
Following code and error might be related, but definitely different function/api/operation call.

Reply:
We don't need to check the contents of `stderr` because we make a binary decision when failures occur: we fail the program and print the reason for the failure if one occurs. However, we do print the stderr when we fail on line 114 so that the end-user/caller can decide how to proceed.

Reply:
It is very unusual the code does not check stderr.

Reply:
@ccfishk could you please propose a code suggestion here? I think I don't understand the concern.

Reply:
if strings.contains(stderr, some-error-keyword(i.e"err")){
  return fmt.Errorf("failed to ")
}
does it make sense to you ?